trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/witness/CorrectnessWitnessExtractor.java

```diff
@@ -27,610 +27,94 @@
-			loopStatements.stream()
-					.map(a -> new ExtractedWitnessInvariant(dwnode.getInvariant(),
-							Collections.singletonList(dwnode.getName()), a, false, false, true))
-					.forEach(a -> rtr.put(a.getRelatedAstNode(), a));
-			iter.remove();
-		}
-		return rtr;
-	}
-
-	private Map<IASTNode, ExtractedWitnessInvariant> tryMatchLoopInvariantsUpwards(final DecoratedWitnessNode dwnode,
-			final Set<MatchedASTNode> loopHeads) {
-		final Map<IASTNode, ExtractedWitnessInvariant> rtr = new HashMap<>();
-		final IASTNode commonParent = findCommonParentStatement(loopHeads);
-		if (commonParent == null) {
-			// if there is no common parent, we cannot match upwards
-			return rtr;
-		}
-		mLogger.info("Matched remaining loop heads upward to common parent of type "
-				+ commonParent.getClass().getSimpleName());
-
-		// check if the common parent or a parent of the common parent is a loop
-		IASTNode currentParent = commonParent;
-		Set<IASTStatement> loopStatements = Collections.emptySet();
-		while (currentParent != null && currentParent instanceof IASTStatement) {
-			loopStatements = CdtASTUtils.findDesiredType(currentParent, mLoopTypes);
-			if (!loopStatements.isEmpty()) {
-				break;
-			}
-			currentParent = currentParent.getParent();
-		}
-		if (loopStatements.isEmpty()) {
-			return rtr;
-		}
-		loopStatements.stream()
-				.map(a -> new ExtractedWitnessInvariant(dwnode.getInvariant(),
-						Collections.singletonList(dwnode.getName()), a, false, false, true))
-				.forEach(a -> rtr.put(a.getRelatedAstNode(), a));
-		loopHeads.clear();
-		return rtr;
-	}
-
-	private static Map<IASTNode, ExtractedWitnessInvariant>
-			extractStatementInvariants(final DecoratedWitnessNode dwnode, final Set<MatchedASTNode> candidateNodes) {
-		final Map<IASTNode, ExtractedWitnessInvariant> rtr = new HashMap<>();
-		candidateNodes.stream()
-				.map(a -> new ExtractedWitnessInvariant(dwnode.getInvariant(),
-						Collections.singletonList(dwnode.getName()), a.getNode(), !a.isIncoming(), a.isIncoming(),
-						false))
-				.forEach(a -> rtr.put(a.getRelatedAstNode(), a));
-		candidateNodes.clear();
-		return rtr;
-	}
-
-	private Map<IASTNode, ExtractedWitnessInvariant> mergeMatchesIfNecessary(
-			final Map<IASTNode, ExtractedWitnessInvariant> mapA, final Map<IASTNode, ExtractedWitnessInvariant> mapB) {
-		if (mapA == null || mapA.isEmpty()) {
-			return mapB;
-		}
-		if (mapB == null || mapB.isEmpty()) {
-			return mapA;
-		}
-
-		final Map<IASTNode, ExtractedWitnessInvariant> rtr = new HashMap<>(mapA.size());
-		for (final Entry<IASTNode, ExtractedWitnessInvariant> entryB : mapB.entrySet()) {
-			final ExtractedWitnessInvariant aWitnessInvariant = mapA.get(entryB.getKey());
-			if (aWitnessInvariant == null) {
-				rtr.put(entryB.getKey(), entryB.getValue());
-			} else {
-				rtr.put(entryB.getKey(), mergeAndWarn(aWitnessInvariant, entryB.getValue()));
+	public HashRelation<IASTNode, IExtractedWitnessEntry> getMatchingWitnessEntries() {

```

Comment:
nitpick: I find "matching" a bit confusing here, as the signature does not tell me what is matched against.

Reply:
I will add some documentation here to help.

Reply:
:+1: What about renaming the method to `getWitnessEntries` ?